Improve utils::hash

deuxksix opened the pull request:
Was working on something else, when I saw the previous recursive hash, which was likely to not properly optimized by the compiler and spend unnecessary cycles, while being simple enough to have a faster implementation.

Several changes:
 - only process first 12 bytes, to avoid accidental infinite loops
 - iterative version is better optimized, often inlined unrolled
 - stricter version for compile-time `hash("string")` usage

Uses the same [djb2 algorithm](http://www.cse.yorku.ca/~oz/hash.html) but process the bytes in a different order than the recursive version for performance, which is incompatible with previous utils::hash (which should not cause problem)

-------------

**Note:** the 12-bytes limit was intentionally set low to enable aggressive optimizations (inline+unroll)

To avoid misuse, an error is raised if a longer string is used, as seen [on that example](https://godbolt.org/z/MnbPM9s16):
```
<source>: In instantiation of 'constexpr uint32_t hash(const char (&)[N]) [with short int N = 14; uint32_t = unsigned int]':
<source>:33:28:   required from here
   33 |     return (hash(h) == hash("magic_string+"));
      |                        ~~~~^~~~~~~~~~~~~~~~~
<source>:22:23: error: static assertion failed: Use hash(s, maxSize) to hash strings longer than 12 bytes!
   22 |     static_assert(N-1 <= 12, "Use hash(s, maxSize) to hash strings longer than 12 bytes!");
      |                   ~~~~^~~~~
<source>:22:23: note: the comparison reduces to '(13 <= 12)'
Compiler returned: 1
```

And of course, the longer version is also `constexpr` optimized, but will be larger to inline and more expensive.

BaptisteHudyma commented on src/system/utils/utils.h:
```diff
@@ -72,10 +72,43 @@ constexpr float map(float x, float in_min, float in_max, float out_min,
+  }
+
+  uint32_t hashAcc = 5381;
+  for (uint16_t I = 0; I + 1 < N; ++I) {
```
`for (uint16_t I = 0; I + 1 < 12; ++I)` to prevent runtime errors

deuxksix replied:
The second implementation of `hash(const char (&s)[N])` is used for compile-time evaluation of constant strings of length known size `N` — for example when `hash("string")` is computed.

This second implementation is equivalent to the first one, but is likely to never be used at runtime, and has been introduced to prevent misuse of the `utils::hash` function (that has a relatively short default `maxSize`).

Marking this as resolved, as there seems to be no runtime errors here.

deuxksix commented:
Momentarily forgot that we need to be careful every time we attempt to use shiny new c++17 things :disappointed: 

-----

Now build works, added a tip for when `build-clean` breaks, also renamed `LMBD_EXPLICIT_CPP17_SUPPORT`

**Note:** now both `hash()` uses the same implementation and there is no longer ambiguity around what they do :+1:

BaptisteHudyma commented:
Thanks for your work !
